rpcs3/rpcs3qt/game_list_frame.cpp

```diff
@@ -2277,86 +2277,132 @@ void game_list_frame::RemoveHDD1Cache(const std::string& base_dir, const std::st
+void game_list_frame::BatchActionBySerials(progress_dialog* pdlg, const std::set<std::string>& serials, QString progressLabel, std::function<bool(const std::string&)> action, std::function<void(u32, u32)> cancel_log, bool refresh_on_finish, std::function<bool()> should_wait_cb)
 {
-	const std::string vsh_path = g_cfg_vfs.get_dev_flash() + "vsh/module/";
-	const bool vsh_exists = game_data.empty() && fs::is_file(vsh_path + "vsh.self");
-	const usz total = !game_data.empty() ? game_data.size() : (m_game_data.size() + (vsh_exists ? 1 : 0));
+	QTimer* timer = new QTimer(this);
 
-	if (total == 0)
-	{
-		QMessageBox::information(this, tr("LLVM Cache Batch Creation"), tr("No titles found"), QMessageBox::Ok);
-		return;
-	}
+	const std::shared_ptr<std::function<bool(int)>> iterate_over_serial = std::make_shared<std::function<bool(int)>>();
 
-	if (!m_gui_settings->GetBootConfirmation(this))
-	{
-		return;
-	}
+	const std::shared_ptr<int> index = std::make_shared<int>(0);
 
-	const QString main_label = tr("Creating all LLVM caches");
+	*iterate_over_serial = [=, this](int index)
+	{
+		if (index == serials.size())
+		{
+			return false;
+		}
 
-	progress_dialog* pdlg = new progress_dialog(tr("LLVM Cache Batch Creation"), main_label, tr("Cancel"), 0, ::narrow<s32>(total), false, this);
-	pdlg->setAutoClose(false);
-	pdlg->setAutoReset(false);
-	pdlg->show();
-	QApplication::processEvents();
+		const auto& serial = *std::next(serials.begin(), index);
+		const int ctr_size = ::narrow<int>(serials);
 
-	u32 created = 0;
+		if (pdlg->wasCanceled() || g_system_progress_canceled)
+		{
+			cancel_log(index, ctr_size);
+			index = ctr_size;
+		}
+		else if (action(serial))
+		{
+			pdlg->SetValue(index + 1);
+		}
 
-	const auto wait_until_compiled = [pdlg]() -> bool
-	{
-		while (!Emu.IsStopped())
+		if (index + 1 == ctr_size)
 		{
-			if (pdlg->wasCanceled())
+			pdlg->setLabelText(progressLabel.arg(index + 1).arg(ctr_size));
+			pdlg->setCancelButtonText(tr("OK"));
+			QApplication::beep();
+
+			if (refresh_on_finish && index)
 			{
-				return false;
+				Refresh(true);
 			}
-			QApplication::processEvents();
+
+			return false;
 		}
+
 		return true;
 	};
 
-	if (vsh_exists)
+	connect(timer, &QTimer::timeout, this, [pdlg, timer, iterate_over_serial, should_wait_cb, index]()
 	{
-		pdlg->setLabelText(tr("%0\nProgress: %1/%2. Compiling caches for VSH...", "Second line after main label").arg(main_label).arg(created).arg(total));
-		QApplication::processEvents();
+		if (should_wait_cb && should_wait_cb())
+		{
+			// Conditions are not met for execution
+			return;
+		}
 
-		if (CreateCPUCaches(vsh_path) && wait_until_compiled())
+		if ((*iterate_over_serial)(*index))
 		{
-			pdlg->SetValue(++created);
+			*index += 1;
 		}
+		else
+		{
+			pdlg->deleteLater();
+			timer->stop();
+			timer->deleteLater();
+		}
+	});
+
+	// Invoked on the next event loop processing iteration
+	timer->start(0);
```

Comment:
using a timer with 0 timeout is disencouraged by Qt.
No idea why you are using this weird method in the first place.
This is spaghetti code at its best.